### PR TITLE
Merge workflows

### LINKED ISSUE #7655: Speed up the new CI jobs introduced in #7653
The idea is to upload artifacts in the same workflow and use them in the dogfooding workflow instead of building anew each time.

That should probably save around 20% of time spent in that whole workflow or a couple percent for the whole CI run (more once we manage to run Windows with more GHCs).

There's TODO in the CI template in that PR. The GitHub Actions docs are at https://docs.github.com/en/actions/guides/storing-workflow-data-as-artifacts

### LINKED ISSUE #7742: Extend new dogfooding ci job to linux and macos
* The motivation is clear: cabal HEAD should be able to install istself including its dependencies. It wuld cover the build of important libs for the ecosystem like network
* The job was initially added for windows by @Mikolaj 

https://github.com/haskell/cabal/blob/69f62284bb46af37d860c82eb0d55506b80ea281/.github/workflows/windows.yml#L72
https://github.com/haskell/cabal/blob/69f62284bb46af37d860c82eb0d55506b80ea281/.github/workflows/windows.yml#L168

* Ideally those jobs should reuse the cabal excutable buils in the previous step. The usual way is upload and download it as a workflow artifact

### PR DESCRIPTION
**Pull request by jneira**
* Continuation of #7952 and #8000
* Goals:
  * Unify the testing workflows for each os, removing the existing duplication
  * Continue using the `validate.sh` script as the primary way of running test suites
  * Make the workflow as uniform as possible for all os's, trying to have the same coverage for all of them
  * Try to keep the job matrix as simpler as possible, using defaults or dedicated steps with branch logic
  * Mark workflow as green, marking known broken tests as such, linking the relevant issues
  * comment all possible caveats, corner cases and less obvious rationales 
  * Closes #7742 
  * Closes #7655

* Bonus
  * validate.sh is usable locally in windows, running it in a msys2 shell
  * each workflow run will have the cabal executables attached as artifacts, so you could download it and reproduce errors locally
  * we will be able to replace all validate jobs as required checks to merge with only one job. And we could change them in the future without touching branch protection rules again

* Continuing the work started in my fork with https://github.com/jneira/cabal/pull/5
  * Succesful workflow here: https://github.com/jneira/cabal/actions/runs/1980092202 

* Caveats
  * You cant restart workflows for os, so a flaky one will make restart the entire workflow
  * We are testing more configurations so times has increased:
    *  with this workflow in my repo : 2h 1m 3s for this one: https://github.com/jneira/cabal/actions/runs/1977079467
        * let see how is going in this one (it has more parallel jobs) 
    *  actual workflows: windows: 1h 50m 15s, macos:  1h 15m 54s and linux  55m 18s for https://github.com/haskell/cabal/pull/7967
  * Also cache size has increased dignificantly, as we have more configurations and we are caching effectively the build dir:
    * Linux 240 Mb *  7 = 1621 Mb
    * Macos 180 Mb * 7 = 1260 Mb
    * Windows 300 Mb * 6 = 1800 Mb
    * ~5 Gb per commit, as github has 10 Gb limit the cache will be invalidated in the third commit of any ongoing pr
      * So the build dir (dist-*) cache will be invalidated very quickly

### CODE REVIEW

File: .github/workflows/validate.yml [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+  validate:
+    name: Validate ${{ matrix.os }} ghc-${{ matrix.ghc }}
+    runs-on: ${{ matrix.os }}
+    outputs:
+      GHC_FOR_RELEASE: ${{ format('["{0}"]', env.GHC_FOR_RELEASE) }}
+    strategy:
+      fail-fast: true
+      matrix:
+        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
+        ghc: ["9.2.1", "9.0.2", "9.0.1", "8.10.7", "8.8.4", "8.6.5", "8.4.4", "8.0.2"]
```
[the remark this answers is not in the document]
Comment (Mikolaj) [reply .github/workflows/validate.yml:41]: Agreed.

**Comment by jneira** [comment]
CI id not picking the new workflow 🤔

**Comment by jneira** [comment]
nothing like a rebase to fix all problems, maybe it was confused due to my fork pr

File: cabal-testsuite/PackageTests/CustomPreProcess/cabal.test.hs [hunk 2]
```diff
@@ -2,12 +2,15 @@ import Test.Cabal.Prelude
 -- Test internal custom preprocessor
 main = cabalTest $ do
     skipUnless "no Cabal for GHC" =<< hasCabalForGhc
+    win <- isWindows
+    ghcsWithMaxPathIssue <- isGhcVersion "<= 8.6.5"
+    expectBrokenIf (win && ghcsWithMaxPathIssue) 6271 $ do
```
[the remark this answers is not in the document]
Comment (jneira) [reply cabal-testsuite/PackageTests/CustomPreProcess/cabal.test.hs]: Fixed, passing  `-w ghc` to validate.sh and making it honour the argument in the cli-suite step

File: .github/workflows/validate.yml [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+      - name: Validate lib-suite
+        run: sh validate.sh -j 2 -w ghc-${{ matrix.ghc }} -v $FLAGS -s lib-suite
+
+      - name: Validate cli-tests
+        if: matrix.cli != 'false'
+        run: sh validate.sh -j 2 -w ghc-${{ matrix.ghc }} -v $FLAGS -s cli-tests
+
+      - name: Validate cli-suite
+        if: matrix.cli != 'false'
+        run: sh validate.sh -j 2 -w ghc -v $FLAGS -s cli-suite
```
Comment (jneira) [inline .github/workflows/validate.yml:154]: oops, i've lost the comment about why using raw ghc intead ghc-x.y.z

Comment (andreabedini) [reply .github/workflows/validate.yml:157]: Should we do the same in the other steps too?
Actually, `validate.sh` would use by default `ghc-8.2.2` ([link](https://github.com/haskell/cabal/blob/906a8bcabc307c6c607cf1b9826a6b329e3ceb7f/validate.sh#L7). If we change it to a generic `ghc` we could avoid passing `-w` everywhere.

File: .github/workflows/validate.yml [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        with:
+          name: cabal-${{ runner.os }}-${{ matrix.ghc }}
+          path: ${{ env.CABAL_EXEC_TAR }}
+
+      - name: Validate lib-tests
+        env:
+          # `rawSystemStdInOut reports text decoding errors`
+          # test does not find ghc without the full path in windows
+          GHCPATH: ${{ steps.setup-haskell.outputs.ghc-exe }}
+        run: sh validate.sh -j 2 -w $GHCPATH -v $FLAGS -s lib-tests
```
Comment (andreabedini) [inline .github/workflows/validate.yml:143]: This is weird. You mean `-w ${{ matrix.ghc }}` would not work here?

**Review by andreabedini — APPROVED** [review]
Looks good to me. Epic work!

File: .github/workflows/validate.yml (hunk 3 above)
Comment (jneira) [reply .github/workflows/validate.yml:157]: hmm make sense, i wanted to touch validate.sh as little as possible but that default is not sensible anymore

File: .github/workflows/validate.yml (hunk 4 above)
Comment (jneira) [reply .github/workflows/validate.yml:143]: it did not work for windows in my tests, will double check and post here or openan issue about

File: cabal-install/tests/UnitTests/Distribution/Client/VCS.hs [hunk 5]
```diff
@@ -48,33 +50,41 @@ import UnitTests.TempTestDir (withTestDir, removeDirectoryRecursiveHack)
 --
 tests :: MTimeChange -> [TestTree]
 tests mtimeChange = map (localOption $ QuickCheckTests 10)
-  [ testGroup "git"
+  [ knownBrokenInWindows "See issue #XXXX" $
```
[the remark this answers is not in the document]
Comment (jneira) [reply cabal-install/tests/UnitTests/Distribution/Client/VCS.hs]: done #8048

File: .github/workflows/validate.yml (hunk 4 above)
Comment (jneira) [reply .github/workflows/validate.yml:142]: Just [double checked](https://github.com/jneira/cabal/runs/5553467288?check_suite_focus=true#step:13:14) ant if fails for all ghc versions with:

```
Unit Tests
  Distribution.Simple.Utils
    rawSystemStdInOut reports text decoding errors:                      FAIL
      Exception: ghc-9.2.1: createProcess: does not exist (No such file or directory)
      Use -p '/rawSystemStdInOut reports text decoding errors/' to rerun this test only.
```

Comment (jneira) [reply .github/workflows/validate.yml:142]: Not sure if it would be related with https://github.com/haskell/cabal/issues/7075 as the path to the ghc executable will be longer than `ghc-x.y.z` for sure

Comment (jneira) [reply .github/workflows/validate.yml:142]: Soooo change `-w ghc-x.y.z` to `-w ghc` seems to workaround also this one :open_mouth: 

https://github.com/jneira/cabal/runs/5553870804?check_suite_focus=true

Maybe the wrapper is interfering with the executable discovering.

So one more reason to switch to raw ghc.
Comments should be moved to validate sh though

**Comment by jneira** [comment]
> You cant restart workflows for os, so a flaky one will make restart the entire workflow

just I ve seen GitHub has added a `rerun jobs from failed`, yay!

**Comment by jneira** [comment]
macOS dogfooding job is failing cause the uploaded tar with the artifact has a folder named `GNUSparseFile.0` with the executable inside:

https://github.com/haskell/cabal/suites/5663836477/artifacts/185963938

That did not happen in previous runs nor in my fork, using the same version of the virtual environment:

https://github.com/jneira/cabal/suites/5665936583/artifacts/185923656

weird

File: .github/workflows/validate.yml (hunk 3 above)
Comment (jneira) [reply .github/workflows/validate.yml]: done

File: .github/workflows/validate.yml [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+  pull_request:
+  release:
+    types:
+      - created
+
+env:
+  # We choose a stable ghc version across all os's
+  # which will be used to do the next release
+  GHC_FOR_RELEASE: '8.10.7'
+  # Ideally the three ghcs should be the used for the release
```
Comment (robx) [inline .github/workflows/validate.yml:27]: I don't understand this comment

File: .github/workflows/validate.yml [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+      - name: print-config using cabal HEAD
+        run: sh validate.sh ${{ env.COMMON_FLAGS }} --with-cabal ./cabal-head/cabal -s print-config
+
+      # We dont use cache to force a build with a fresh store dir and build dir
+      # This way we check cabal can build all its dependencies
+      - name: Build using cabal HEAD
+        run: sh validate.sh ${{ env.COMMON_FLAGS }} --with-cabal ./cabal-head/cabal -s build
+
+  # We use this job as a resume of the workflow
```
Comment (robx) [inline .github/workflows/validate.yml:265]: I think "summary" rather than "resume" would be clearer .

File: cabal-install/tests/UnitTests/Distribution/Client/VCS.hs [hunk 8]
```diff
@@ -48,33 +50,41 @@ import UnitTests.TempTestDir (withTestDir, removeDirectoryRecursiveHack)
 --
 tests :: MTimeChange -> [TestTree]
 tests mtimeChange = map (localOption $ QuickCheckTests 10)
-  [ testGroup "git"
+  [ knownBrokenInWindows "See issue #8048" $
```
Comment (robx) [inline cabal-install/tests/UnitTests/Distribution/Client/VCS.hs:53]: It might be worth skipping these instead, because I think they're quite slow due to lots of disk IO.

File: cabal-testsuite/src/Test/Cabal/Server.hs [hunk 9]
```diff
@@ -254,7 +254,8 @@ initServer s0 = do
     pid <- withProcessHandle (serverProcessHandle s0) $ \ph ->
               case ph of
                   OpenHandle x   -> fmap show (Win32.getProcessId x)
-                  ClosedHandle  _ -> return (serverProcessId s0)
+                  -- TODO: handle OpenExtHandle?
+                  _              -> return (serverProcessId s0)
```
Comment (robx) [inline cabal-testsuite/src/Test/Cabal/Server.hs:258]: Looks like the `#ifdef` could be reduced to just the `OpenHandle` line now.

**Review by robx — APPROVED** [review]
Generally looks good to me, left a couple minor comments.

No opinion on the general issues (total running time, caching).

File: .github/workflows/validate.yml (hunk 7 above)
Comment (jneira) [reply .github/workflows/validate.yml:265]:
```suggestion
  # We use this job as a summary of the workflow
```

File: .github/workflows/validate.yml (hunk 6 above)
Comment (jneira) [reply .github/workflows/validate.yml:27]: nither me 😄

File: cabal-install/tests/UnitTests/Distribution/Client/VCS.hs (hunk 8 above)
Comment (jneira) [reply cabal-install/tests/UnitTests/Distribution/Client/VCS.hs:53]: fair

File: cabal-testsuite/src/Test/Cabal/Server.hs (hunk 9 above)
Comment (jneira) [reply cabal-testsuite/src/Test/Cabal/Server.hs:258]: oh yeah, good catch

**Comment by jneira** [comment]
ok tar definitely is related, I executed `file ./path/to/cabal` before tar and it reported `cabal: Mach-O 64-bit executable x86_64`
After untar the same file it reported `cabal: data`

**Comment by robx** [comment]
> ok tar definitely is related, I executed `file ./path/to/cabal` before tar and it reported `cabal: Mach-O 64-bit executable x86_64` After untar the same file it reported `cabal: data`

Just to note that I can confirm this locally: Downloaded the artifact from the bottom here https://github.com/haskell/cabal/actions/runs/2000402197, and it shows the same when I unzip and untar it. Let me know if you'd like me to test something on macOS.

**Comment by jneira** [comment]
> Let me know if you'd like me to test something on macOS.

many thanks for reproduce and report it :slightly_smiling_face:

**Comment by jneira** [comment]
Related: https://github.com/actions/virtual-environments/issues/2619
Or i hope so, as use gnutar seems to workaround the issue

**Comment by Mikolaj** [comment]
Isn't there an option to tar for compatibility with gnutar?

**Comment by jneira** [comment]
> Isn't there an option to tar for compatibility with gnutar?

i've decided to try the other workarounf dirst: call `sudo /usr/sbin/purge` to clean cache before do the tarball, as it does need any additional installation

**Comment by jneira** [comment]
`PackageTests\InternalLibraries\Haddock\haddock.test.hs` is flaky on windows:

```

<command line>: does not exist: C:\Users\RUNNER~1\AppData\Local\Temp\ghcB217.c


stderr:
*** Exception: Command "D:/a/cabal/cabal/dist-newstyle-validate-ghc-8.6.5/build/x86_64-windows/ghc-8.6.5/cabal-testsuite-3\build\setup\setup" "configure" "-vverbose +markoutput +nowrap" "--distdir" "haddock.dist\work\dist" "--global" "--with-ghc" "C:\ProgramData\chocolatey\lib\ghc.8.6.5\tools\ghc-8.6.5\bin\ghc.exe" "--with-haddock" "C:\ProgramData\chocolatey\lib\ghc.8.6.5\tools\ghc-8.6.5\bin\haddock.exe" "--enable-deterministic" "--prefix=D:\a\cabal\cabal\cabal-testsuite\PackageTests\InternalLibraries\Haddock\haddock.dist\usr" "--package-db=clear" "--package-db=global" "--package-db=C:\sr\ghc-8.6.5\package.db" "--package-db=D:\a\cabal\cabal\dist-newstyle-validate-ghc-8.6.5\packagedb\ghc-8.6.5" "--package-db=D:\a\cabal\cabal\cabal-testsuite\PackageTests\InternalLibraries\Haddock\haddock.dist\packagedb" failed.
```

We should consider ignore it

**Comment by jneira** [comment]
> > Isn't there an option to tar for compatibility with gnutar?
> 
> i've decided to try the other workaround suggested in the issue: call `sudo /usr/sbin/purge` to clean cache before do the tarball, as it does need any additional installation

it did the trick!

@Mikolaj @phadej i will wait to merge some hours just in case you have the opportunity of take a final look
After that we will have to manually merge it and change again required checks and rebase other prs. But it will be the last time, as we will use the final post job as the unique require check for the workflow and it likely will not be changed

**Review by Mikolaj — APPROVED** [review]
Great job.

We should rebrand cabal as CI consultancy and we'd start bringing in profit for HF.
